### PR TITLE
feat: ADR-049 Phases 5-6 — every front end consumes one resolved config

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Joins ADR-049 Phase 1's canonical resolver to the front ends that were each reconstructing their own narrower configuration object. `compare`, the MCP `abi_compare` tool, `scan --against`, and the `run_scan` Python API now every one resolve a single `CompatibilityEvaluationConfig` carrying real D7 per-field provenance — closing §12's Definition-of-Done item 2. Adds the unsuppressible contract-coverage ledger (item 6), the `--contract public|exports|all` domain selector (Phase 6), and the platform corpus that measures it.

**Advisory throughout.** No verdict, finding set, or exit code changes. Phase 7 owns the default flip.

> **Note:** this description was rewritten at `e1387c2`. The original was written when the PR was Phase 5 only, and its "still open" list named four items that this same PR went on to complete — it no longer described the diff.

## Motivation

Phase 1 built the resolver and Phase 4 persisted an `evaluation_context`, but nothing joined them. A core verb sees *values*, not the inputs that chose them, so every stated value was recorded at the `api_request` layer — honest, but useless to an audit consumer asking *which* layer selected a value, which is the whole point of a D7 receipt.

Wiring it in exposed a real bug: `--profile ci-gate` injects `exit_code_scheme: severity` without stamping a parameter source, so a receipt resolved without knowledge of the profile answered `legacy` for a run that really scored under the severity-aware scheme. A wrong value, not merely an unnamed source.

## What's in it

**Front ends → one resolved config**
- `cli_compare_receipt.py` — resolves from raw CLI values, the parameters Click reports as *typed*, the discovered `.abicheck.yml`, a selected `--profile` (D7's `run_profile` tier), and already-loaded policy/suppression documents (passed in, never re-read, so one content's digest can't be paired with another's rules).
- `mcp_compare_receipt.py` — the same for `abi_compare`, which previously patched its own gate. The receipt states only what that tool *has*: it takes no scope/contract/public-symbol parameter, so those resolve as built-in defaults.
- `cli_scan_receipt.py` + `service_scan.py` — `scan --against` and `run_scan`. Reuses `compare_cli_inputs` rather than a second normalizer, because §6.4's gate is that the two commands agree and two normalizers is how they'd stop.
- `contract_context.with_resolved_config` — the install seam. Keeps `contract.overlays`/`surface.explicit_scope` recovered from the evidence ledger whenever an overlay was observed: an observation of what ran outranks a resolution of what was asked for.

**The gate is a deliberate split** — values from `resolve_compare_config` (what actually scored the run), provenance from the canonical resolver. Sound only while the two agree, so a parametrized parity matrix compares the scheme and all four severity levels.

**Unsuppressible coverage ledger** (`contract_coverage_ledger.py`) — derives `contract_coverage_failures` per selected domain from the observed provider records. Unsuppressibility is *structural*: a `CoverageFailure` is not a `Change`, so `checker._filter_suppressed_changes` cannot see one. Emitted in JSON (report schema 2.26), SARIF (`toolExecutionNotifications`), and JUnit (an error suite, never passing tests), none of which touch the run's gate.

**Phase 6** — `--contract` selects the evidence domain the shadow evaluator judges against; `scripts/contract_platform_corpus.py` adds ELF/PE/Mach-O/versioned/extern-C/stripped lanes with per-lane unresolved-rate reporting.

**Receipt honesty, as a check** — `unstatable_selectors()` reports any provenance hop naming an input its own selector layer cannot state. Review found that defect four separate times; the cross-front-end equality gate structurally cannot catch it, because it normalizes option spellings away on purpose.

## Known gaps

- **`packs` parity axis is open.** Nothing selects packs: `pack_paths` is only ever `()`. A `--pack` flag was added and then *removed* in this PR rather than shipped, because it resolved packs into the receipt without ever feeding `policy.overrides` into the verdict path — configuration that did not configure. Making it real needs its own slice plus re-verification against the FP-rate and tier-accuracy gates.
- Phase 7 (default flip, and wiring `coverage_exit_contribution` to a real exit code) is untouched, gated on release notes and a migration window.

## Checklist

- [x] Tests added/updated — ~3,300 lines across 9 files (config receipts for all four front ends, the coverage ledger, the §6.4 parity matrix, the platform corpus, schema-drift pinning)
- [x] Changelog fragments added (`scriv create`)
- [x] Docs updated — plan Phase 5/6 sections, `docs/reference/compatibility-evaluation-config.md`, `AGENTS.md` module map, regenerated CLI/API references
- [x] `pytest tests/ -m "not integration and not libabigail and not abicc and not slow and not golden"` — 22702 passed; `check_ai_readiness.py` and `check_docs_contract.py` 0 errors
- [x] No new `mypy`/`ruff` errors

### CODE REVIEW

File: abicheck/cli_compare_helpers.py [hunk 1]
```diff
@@ -1693,17 +1693,58 @@ def run_compare(
...
+            project_cfg,
+            # The raw CLI values, not the already-merged ones several of
+            # these locals were overwritten with above: the resolver's whole
+            # job is to merge them itself, from each layer's own inputs, and
+            # handing it a pre-merged value would make every field look
+            # CLI-stated.
+            params={
+                "contract_mode": contract_mode,
+                "scope_public_headers": scope_public_headers,
+                "policy": policy,
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_compare_helpers.py:1734]: Real, and fixed in 826f922 — reproduced first: `compare --required-symbol api_b --contract-evaluation` did persist `policy.base: strict_abi` for a run scored under `plugin_abi`.

Adding `policy` to `typed` would have been the wrong fix, since the receipt would then claim a `--policy` the user never passed. `ExplicitCompatibilityInputs` gained `policy_base_option` instead: the derived value is read as stated regardless of the typed set, and the candidate records `--required-symbol` as its selector. It keeps the `LEGACY_ALIAS` slot, so an explicit `--policy-file` still outranks it — matching what `effective_policy` does live.

Two tests: `test_required_symbol_contract_records_the_policy_it_really_used` (value plus the named selector) and `test_typed_policy_still_wins_over_the_required_symbol_default` (the derivation is gated on the flag being untouched, so a typed `--policy` must still be reported as the selector).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/contract_context.py [hunk 2]
```diff
@@ -297,6 +301,67 @@ def with_field_provenance(
...
+
+    observed = context.evaluation_context.resolved_config
+    if observed.contract.overlays:
+        provenance = dict(config.provenance)
+        for field in (CONTRACT_OVERLAYS_FIELD, EXPLICIT_SCOPE_FIELD):
+            entry = observed.provenance.get(field)
+            if entry is None:
+                provenance.pop(field, None)
+            else:
+                provenance[field] = entry
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/contract_context.py]: Right, and fixed in 826f922. The sharpening is the useful part: value and receipt answer different questions, so they should not have followed one rule. The value stays the ledger's — it names what actually applied, including a `--post-manifest` scope the front end cannot model — while the provenance now keeps the resolver's entry (layer, option, path, digest) and appends the observed hop, so the selection *and* its application are both recorded. The core's entry stands alone only when the front end resolved no scope at all.

`_merged_scope_provenance` handles the two one-sided cases and drops a duplicate hop, so a re-applied merge cannot report one selection as two. Tests: `test_a_symbols_list_file_is_still_named_in_the_receipt` (end-to-end: the path and `--public-symbols-list` appear alongside the `overlays` hop, value still the ledger's) plus two unit-level edge cases for a context whose overlay entries are missing or already merged.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_compare_helpers.py [hunk 3]
```diff
@@ -1234,8 +1234,15 @@ def run_compare(
...
     # user explicitly picked one -- an explicit --policy always wins (ADR-043).
+    # Recorded, not just applied: the ADR-049 receipt has to name whatever
+    # really selected `policy.base`, and this value was chosen by a typed
+    # `--required-symbol` rather than by `--policy` or a built-in default
+    # (Codex review, fresh evidence -- the receipt claimed `strict_abi` for a
+    # run that used `plugin_abi`).
+    policy_selected_by: str | None = None
     if required_symbols and ctx.get_parameter_source("policy") != click.core.ParameterSource.COMMANDLINE:
         policy = "plugin_abi"
+        policy_selected_by = "--required-symbol"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_compare_helpers.py]: Correct — fixed in fa554e2. Whichever option really supplied the contract is what the receipt names now: `--required-symbols` whenever a file was given (it is a real option the user passed either way, and the one carrying something to audit), `--required-symbol` for an inline-only run.

The path and a sha256 travel with it. `_load_required_symbols` now returns the digest from the same read that parses the symbols, so the receipt identifies the exact list that selected `plugin_abi` rather than just the flag — recorded as `source_sha256`, since it identifies the *source*, while the value is a file-less built-in policy identity with a digest of its own. Test: `test_a_symbols_file_contract_names_the_file_option_and_the_file`, asserting the option, the path, and a digest matching `sha256(file bytes)`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_compare_receipt.py [hunk 4]
```diff
@@ -61,118 +90,174 @@
...
+        explicit=compare_cli_inputs(
+            params,
+            explicit_parameters=typed,
+            policy_file=policy_file,
+            suppression=_suppression_source(suppression, suppress_path),
+            policy_base_option=policy_option,
+        ),
+        project=ProjectCompatibilityInputs.from_build_config(
+            project_cfg, path=project_path
+        ),
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_compare_receipt.py:225]: Agreed — fixed in fa554e2, and the interesting part was *where* the digest may be taken.

`ProjectCompatibilityInputs.sha256`'s own docstring refuses to compute it from the path, for the pairing reason a second read reintroduces (one revision's digest against another's values), and asks a caller that read the file to pass it. So the digest now comes from the read that parses the config: `load_build_config_with_digest` returns both, over raw bytes rather than decoded text, so a CRLF file matches what is on disk instead of its LF translation. `_resolve_compare_config` threads it to `ProjectCompatibilityInputs.from_build_config`.

It landed in a new leaf, `buildsource/build_config_io.py`, rather than next to `load_build_config`: `inline.py` is at its 2000-line hard cap, and the dependency runs one way (the new module imports `BuildConfig`; `inline` does not import it back), so no cycle forms and every values-only caller keeps `load_build_config` unchanged. Test: `test_project_config_provenance_carries_its_digest`, asserting the recorded hop's digest equals `sha256(.abicheck.yml bytes)`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/buildsource/build_config_io.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    there was no file to read -- the same all-defaults case
+    ``load_build_config`` tolerates.
+
+    Raises :class:`ValueError` for an unreadable or malformed document, with
+    the same message shape ``load_build_config`` raises, so a caller can
+    handle either identically.
+    """
+    if not path.is_file():
+        return BuildConfig(), None
+    import yaml  # hard dep; imported out of the try so the except can name it
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/buildsource/build_config_io.py:64]: Right, and it was already failing CI — the `ai-readiness` job went red on fa554e2 for exactly this. Fixed in dc31adc by adding `abicheck.buildsource.build_config_io` to the existing stubless-PyYAML override block, not an inline ignore.

Worth noting for anyone reading the thread: it also explains a local symptom I'd written off as environmental. `mypy abicheck/` here reports the same `import-untyped` for several already-overridden modules, so the new module's error looked like more of the same noise rather than the one real regression in it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/contract_context.py [hunk 6]
```diff
@@ -297,6 +301,113 @@ def with_field_provenance(
...
+                config.surface, explicit_scope=observed.surface.explicit_scope
+            ),
+            provenance=provenance,
+        )
+    return replace(
+        context,
+        evaluation_context=replace(context.evaluation_context, resolved_config=config),
+    )
+
+
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/contract_context.py:410]: Good catch — "no front-end input at all" was wrong, and I wrote that claim into the docstring. `CONTRACT_OVERLAYS_FIELD` is routable from a `kind: contract` pack, so both sides can state this field. Fixed in dc31adc.

Taking your first option, with one addition: the values are a **union**, not a replacement. A pack assigns overlay *selectors* while the ledger observes *provider ids* — two different sets, and both were genuinely in effect, so keeping only one loses a real selection as well as its receipt. Provenance now goes through the same merge `surface.explicit_scope` already used (renamed `_merged_overlay_provenance`, since it serves both): the stated entry survives and the observed hop is appended, and the observed entry stands alone when the front end stated nothing.

Not treating a difference as a mismatch to detect, deliberately: disjoint values are the *expected* shape here, not a conflict — genuine same-tier conflicts are already the resolver's job (`FieldResolutionError`, mapped to exit 64 by this front end).

Test: `test_a_stated_overlay_selector_survives_alongside_the_observed_one` — a pack-style `ffi` overlay beside an observed `forced_public_symbols`, asserting the union and that both hops are recorded.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_compare_receipt.py [hunk 7]
```diff
@@ -61,118 +90,201 @@
...
+    # here and re-contributed at the `run_profile` tier below, which is both
+    # the honest source and the precedence D7 gives it.
+    injected = (run_profile or {}).get("injected") or {}
+    params = {
+        name: (None if name in injected else value) for name, value in params.items()
+    }
+    return resolve_compatibility_evaluation_config(
+        front_end=FrontEnd.CLI,
+        explicit=compare_cli_inputs(
+            params,
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_compare_receipt.py:214]: Right — and the deletion case is the sharper half: a file removed after the comparison started would fail an *otherwise finished* run during receipt generation, which is worse than a stale digest. Fixed in bef6c4f, consistent with the other four sources as you say.

`resolve_force_public_scope` (`cli_helpers_compare.py`) does the one read both consumers share; `collect_force_public_symbols` takes `already_read=`, and `compare_cli_inputs` takes `public_symbols_list=` — the same "pass what you already read" affordance `policy_file`/`suppression` already had. Test: `test_a_symbols_list_deleted_mid_run_does_not_fail_the_comparison`, which deletes the file at the moment the receipt is recorded and asserts the run still completes with the path recorded.

Two cleanups fell out. `with_field_provenance` turned out to have lost its last caller when the resolver took over `contract.mode` provenance — deleted rather than left as an uncalled seam. And `cli_compare_helpers.py` crossed its 2000-line cap again, which the helper's new home resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_compare_helpers.py [hunk 8]
```diff
@@ -1234,13 +1256,34 @@ def run_compare(
...
+    # audit.
+    policy_selected_by: str | None = None
+    policy_selected_path: Path | None = None
+    policy_selected_sha: str | None = None
     if required_symbols and ctx.get_parameter_source("policy") != click.core.ParameterSource.COMMANDLINE:
         policy = "plugin_abi"
+        if required_symbols_file is not None:
+            policy_selected_by = "--required-symbols"
+            policy_selected_path = required_symbols_file
+            policy_selected_sha = required_symbols_sha
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_compare_helpers.py]: Right — my rule was still a claim about what was *passed*, not what contributed, and the mixed case exposes the difference. Fixed in 87fd9dd.

`load_required_symbols` now returns what the file itself contributed alongside the combined contract and the digest, all from the one read; the file form is named only when that contribution is non-empty. Took your second option rather than "record all contributing selectors": with either form non-empty the derived `plugin_abi` is reproducible from that one selector, so naming a real contributor is sufficient — what was wrong was naming one that contributed nothing. When both contribute, the file form still wins, since it is then equally true and the one carrying a path and digest to audit.

Test: `test_a_symbols_file_that_contributed_nothing_is_not_the_selector` — exactly your `--required-symbol api_b --required-symbols empty.txt`, asserting the inline option is named and no path is recorded.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_build_config_io.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+from __future__ import annotations
+
+import hashlib
+
+import pytest
+
+from abicheck.buildsource.build_config_io import load_build_config_with_digest
+
+
+def _sha(path) -> str:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_build_config_io.py]: Done in 87fd9dd — `_sha(path: Path)` and every `tmp_path: Path`, no `-> None` added.

Closing out the other two from the same review here, since they were posted as nitpicks rather than threads:

- **`resolve_force_public_scope` types** — agreed and fixed: `str | Path | None` in, `tuple[set[str], PublicSymbolsList | None]` out. `PublicSymbolsList` goes in the existing `TYPE_CHECKING` block, so the annotation is concrete without a runtime import; the construction still imports locally.
- **Plan-doc heading** — declining, with a reason. `public-contract-default.md` uses one convention throughout for this: `**Updated (date): …**` paragraphs appended under the phase they belong to, roughly forty of them by now. A differently-shaped heading for this one entry would break the file's own structure rather than clarify it, and the topic registry already routes readers to `reference/compatibility-evaluation-config.md` — which is where I put the user-facing precedence and the `run_profile` tier, precisely so the plan entry stays a rollout record.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Review round addressed — four of the five taken, one declined with reasoning.

**Taken:**

1. **`schemas/__init__.py` + the plan note — the byte-identity claim was wrong.** Correct on both counts: `detectors` is emitted regardless of `--contract-evaluation`, and `scan_schema_version` itself changes. Both now state exactly what a run *without* the flag changes relative to 1.5 — the version marker, `finding_id`, and `detectors` — with the contract keys still called out as flag-only.

2. **`service_scan.run_scan` — fail fast on `contract_mode` without `contract_evaluation`.** The premise needs one correction: it was *not* silently ignored. `service.compare_snapshots` calls `_validate_contract_mode` at the Tier-2 boundary and raises `ValidationError`, verified directly:

   ```
   >>> compare_snapshots(old, new, contract_mode="exports", contract_evaluation=False)
   ValidationError: contract_mode requires contract_evaluation: ...
   ```

   The real defect is *when* it fires: reaching that check means a full scan runs first and then fails, where `scan_cmd` rejects the same request before any work. `run_scan` now applies the identical rule up front, reusing `_validate_contract_mode` rather than restating it. Test asserts `run_scan_core` was never called.

3. **Truncation assertion in the parity matrix.** Taken — `assert not scan["diff"].get("findings_truncated")` before the set comparison, so a future fixture past `_MAX_BASELINE_FINDINGS` fails by name instead of as an opaque set diff.

4. **Defensive enum extraction in `_add_contract_fields`.** Taken. `_baseline_finding_dicts`' documented contract is that it stays safe against the lightweight fakes the scan tests and sibling modules build, and requiring a real enum here broke that for exactly those callers. Both fields now read through `getattr(..., "value", str(...))`, matching how `kind` is already read, with a regression test that projects a fully string-valued stub.

**Declined:**

5. **Moving `mixed_pair` to `conftest.py`.** The guideline is about shared fixtures; this one is used by a single test class in a single module, alongside four other module-local fixtures that already live there (`runner`, `old_snap`, `new_snap_breaking`, `suppress_bar`). Promoting it to the root `conftest.py` would make it visible to ~180 unrelated test modules to serve one, which is the cost the guideline exists to avoid rather than a case of it. Happy to revisit if a second module needs the same pair.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Both Codex P2 findings confirmed and fixed in `c642e6e`.

**1. Overridden MCP policy reaching the resolver** — real, and the failure mode was as described. `abi_compare` validates `policy` only when no `policy_file` is given, so an unknown policy alongside a valid file is an *accepted* request whose comparison completes under the file's base. Reproduced:

```
>>> resolve_tool_config(policy="not_a_real_policy", policy_file=PolicyFile(base_policy="sdk_vendor"))
ValueError: unknown base policy 'not_a_real_policy': choose from [...]
```

A receipt turning a finished comparison into an error response is the one thing a receipt must never do. The ignored value is now dropped rather than repaired — it chose nothing, so naming it in the receipt would be false either way, and the resolver reads the base off `policy_file` exactly as the comparison did. With no file present it still passes through unchanged, so a genuinely unknown policy fails loudly at the tool's own validation. Four regression tests, including one asserting the completed comparison survives.

**2. Mach-O export spellings** — also real, and the consequence was already visible in the output I'd committed without attributing it correctly. Confirmed against `macho_metadata.py`: both the export-trie walk and the symbol-table fallback strip the leading underscore before constructing a `MachoExport`, so `__Z...` is a spelling no real snapshot can hold. The fixture's phantom export showed up as `macho` 1 resolved / 1 unresolved under `exports` — in a lane whose entire purpose is demonstrating resolution from an observed table. Normalized to `_Z...`; the corpus drops from 56 findings to 55 and the lane is now 0 unresolved in every domain. The lane test additionally asserts `unresolved == 0`, not just `resolved > 0`, so the same class of fixture error fails rather than degrading a number.

---

`c642e6e` also lands the last open Phase 5 item, Definition-of-done #6's second half — the unsuppressible contract-coverage ledger (`contract_coverage_failures` / `contract_coverage_exit_contribution`, report schema 2.26). Unsuppressibility is structural rather than a rule: a coverage failure is not a `Change`, so `checker._filter_suppressed_changes` cannot see one, and `suppression_reaches_coverage_failures()` is the executable proof of that (a wildcard `symbol: '.*'` rule still reaches nothing) rather than its enforcement.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Both Codex findings accepted. The P1 is the important one and it was right.

**P1 — `--pack` selected packs that never reached the engine.** Confirmed: assignments landed in the resolved configuration and the persisted receipt, but `compare_snapshots` never saw them, so a `kind: policy` pack overriding `func_removed` would have been *recorded as active configuration* while leaving the verdict and exit code untouched — and without `--contract-evaluation` the manifest was not loaded at all, so a malformed one was accepted silently.

Worth naming why my own tests missed it: they asserted the two commands *resolve* packs identically and never that a pack changes a result. A flag that does nothing satisfies that assertion perfectly.

I removed the flag rather than patching around it. Exposing configuration that does not configure is worse than not exposing it, and making packs real is its own slice — feeding `policy.overrides` into the verdict path (the `PolicyFile`-shaped overrides `compare_snapshots` already consumes), deciding D8 precedence against an explicit `--policy-file`, and re-verifying against the FP-rate and tier-accuracy gates. Not a drive-by extension of a receipt change.

Consequence I'm stating plainly rather than burying: **the `packs` axis of the Phase 5 Gate is back to open.** The plan note now says so instead of claiming it closed. The `cli.py`/`cli_options.py` hard-limit split the flag forced is kept — `cli_contract_options.py` is a real improvement independent of packs.

**P2 — scan's receipt claimed gate settings the scan discards.** Also correct. A `.abicheck.yml` with `severity.preset: info-only` was recorded in the persisted context with every level at `info`, while the scan still exited 4 on a breaking diff, because a scan's exit follows its compatibility verdict directly and it never consults those keys. `resolve_scan_config` now blanks the project's gate fields before resolving. Blanked rather than omitted: they still resolve, to their built-in defaults, which is the true statement — nothing selected them for this run. Same answer the MCP receipt gives for parameters `abi_compare` does not have.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Both P2s accepted — each is the same defect one surface over from where I wired it, which is a pattern worth naming.

**`to_junit_xml_multi` never emitted the coverage suite.** Correct. I wired `_append_coverage_suite` into the single-result renderer only, so a multi-library document reported `errors="0"` with no `abicheck.contract_coverage` suite at all — letting a consumer read an uncheckable comparison as clean, which is exactly the failure mode the suite exists to prevent. Now called per result (each library carries its own contract context, so one uncheckable comparison can sit beside several closed ones) with its errors rolled into the document totals. Test asserts both the nested suite and the root `errors` count.

**`run_scan` didn't resolve a config.** Also correct, and the same shape as the two receipt findings before it: the CLI path was wired and the Python API path was left behind, so a `run_scan(ScanRequest(..., contract_evaluation=True))` persisted `checker.compare`'s argument-shaped reconstruction — keeping `GateConfig`'s defaults and claiming the `severity` scheme while `run_scan` computed its 0/2/4 exit straight from the compatibility verdict. `_scan_request_config` now resolves one, reading every field as *stated* (a typed request has no "unset", so a caller constructing one chose those values — the rule `compare_request_inputs` already follows) and blanking the gate fields for the same reason the CLI blanks the project config's.

That's four findings in this PR of the form "a receipt claims something the run did not do", across four surfaces: `compare`, `abi_compare`, `scan --against`, and now `run_scan`. The underlying cause each time was wiring one front end and not checking the others. `cross_front_end_differences` catches divergence between *resolved configs*, but nothing catches a front end that never resolves one — worth a follow-up gate rather than a fifth round of the same review finding.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Both Codex findings on `efbfbf4` confirmed and fixed in `7c785af`.

- `service_scan.py` — `resolve_scan_config` hard-coded `FrontEnd.CLI`, so `run_scan(ScanRequest(...))` recorded `policy`/`scope_public` as `--policy`/`--scope-public-headers`. Now takes a `front_end` argument, defaulting to the CLI; `_scan_request_config` passes `FrontEnd.API`. Only the spelling and layer change — the normalization stays shared with `compare` on purpose, since §6.4's gate is that the two agree.
- `mcp_compare_receipt.py` — the resolver's `spell()` helper already exists for this, and the two severity candidate sites (preset and the four category overrides) were built with hard-coded `"--severity-*"` strings and bypassed it. Fixed at those two sites, so every front end benefits rather than just this one.

The more useful part: this is the **third and fourth** instance of the same defect (the first motivated `spell()` itself), so it now gets a check rather than another fix. `unstatable_selectors(config)` reports any `api_request` hop labelled with a CLI flag.

Worth stating why the existing gate missed all four: `cross_front_end_differences` runs everything through `_normalized_provenance`, which drops option spellings **on purpose** — the same semantic input is legitimately spelled differently per front end, and that normalization is what makes the equality gate meaningful at all. It is therefore structurally blind to a hop naming a flag nobody could have typed. That is a separate invariant, so it needed a separate check, not a stricter setting of the existing one.

Deliberately one-directional: a CLI hop carrying a bare field name is not flagged, because several CLI inputs (a project-config key, a composed scope) genuinely have no flag.

One correction to the first finding's framing, since it affects what the fix could assert: `policy.base` resolves at `legacy_alias` on *every* front end — `--policy`/`policy` is D7's documented legacy alias for the base — so the layer there says nothing about which front end stated it. The option spelling was the real defect and is what changed; the layer assertion in the new test is on `contract.mode`, which does resolve at the request tier.

Local: 22689 passed, `ai-readiness` and `docs-contract` 0 errors, ruff/mypy clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Worked through the 8 CodeRabbit comments plus the third Codex P2. Six fixed in `d917ce2`, two declined with reasoning below.

**Codex P2 (`service_scan.py`) — correct, and it caught that my previous fix was half a fix.** Resolving at `FrontEnd.API` fixed the front end but not the *namespace*: the default API spelling is `CompareRequest`'s, and a `ScanRequest` calls the same three inputs `scope_to_public_surface`/`policy_file`/`suppression`. One unreplayable spelling replaced another. The resolver now takes an `api_spellings` remap (`4354349`).

That also showed `unstatable_selectors` was too weak — "not a CLI flag" passes for any plausible identifier. Given `request_type=`, it now checks every caller-stated hop against that dataclass's real fields, and covers `legacy_alias` as well as `api_request`: the reported `scope_public` hop sat at the *former*, so a check limited to the request tier would have passed the defect it was written for. Verified by reproducing the pre-fix state — it reports both offenders.

**Fixed from CodeRabbit:**

- **JUnit library identity** — real. Suite name and testcase classname are now qualified; regression test uses two libraries with identical provider/side failures.
- **`ValidationError` for resolver failures in `run_scan`** — real, and slightly wider than described. Beyond the three named exceptions, an unknown `ScanRequest.policy` raises a *plain* `ValueError` from `builtin_policy_identity` and leaked the same way; the CLI can't reach that case because `--policy` is a `click.Choice`, but a typed request's `policy` is a free string. Catching `ValueError` broadly (plus `PackManifestError`), mapped inside the helper so a second call site can't reintroduce it.
- **Schema `required`** — confirmed against `to_dict()`; `status` and `completeness` added to both copies.
- **`CASE_LANE` import-time guard** and **deriving the lane list from `CASE_LANE`** — both real drift risks, taken as suggested.
- **Blind `pytest.raises(Exception)`** — right about B017 in substance, but the suggested fix would have failed: that path raises a plain `ValueError` from `builtin_policy_identity`, not `FieldResolutionError`/`PackManifestError`. Narrowed to `ValueError` with a `match=`.
- **Resolved-value assertion and test annotations** — taken.

**Declined:**

- **MD046 fenced → indented** — the repo has no markdownlint config, and that page already contains 10 fenced blocks. Converting only the new one would make it the odd block out. This is CodeRabbit's default style, not the repo's.
- **Hand-copied versions in `plans/public-contract-default.md`** — the rule targets a live fact mirrored out of its owner. These are prose in an implementation *record* ("`REPORT_SCHEMA_VERSION` went to `2.26` for the two additive keys"), which stays true when the version later moves; `contribute/plans/` is treated elsewhere in `docs/AGENTS.md` as an inherently historical record for that reason. Linking would show the *current* value, making the record less accurate, not more.

**Docstring-coverage pre-merge check:** not a repo gate. `ai-readiness` has no docstring check, the figure is dominated by test methods, and `tests/CLAUDE.md` sets no such requirement. The production code here is documented.

Local: 22698 passed, `ai-readiness` and `docs-contract` 0 errors, ruff clean, mypy clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Codex's P2 on `mcp_server.py` is correct, and it caught something my own documentation actively denied. Addressed in `9321304`.

**The finding.** `abi_compare` takes `used_by` and `required_symbols`; both are authoritative — the scoping pass rewrites the verdict and exit code from them — but neither reaches `resolve_tool_config`. Meanwhile `docs/reference/compatibility-evaluation-config.md` stated the tool "takes no scope, contract-mode, public-symbol, or exit-code-scheme parameter, so those resolve as `built_in_default` … which is the accurate answer, not a missing claim." For the scope dimension that was simply false.

**It's wider than described.** Codex framed it as the receipt not identifying *which* applications scoped the run. Checking empirically, it doesn't record that a scope applied **at all**: `"scoped"` isn't a value `GateConfig` accepts, so `record_resolved_config` stores the *underlying* scheme instead. A scoped run's `resolved_config` compares equal to an unscoped one's — `gate` and `surface` both.

**Not closing it here, and why.** `policy_base_option` is the CLI's vehicle for `--required-symbol`, and it works there only because the contract switches an untouched `--policy` to `plugin_abi` (ADR-043). The MCP path performs no such switch, so that mechanism has nothing to hang on. And `used_by` is modelled by *neither* front end — `compare --used-by` has the identical gap. A real fix needs a new typed field plus an identity scheme for application binaries (path + digest, resolved once and shared with the comparison that already read them). That's a scoped design, on the same footing as the still-open `packs` axis — not a parameter to thread through.

Recorded in three places so it stays known rather than being rediscovered: the reference (as a warning admonition), `mcp_compare_receipt.resolve_tool_config`'s docstring, and a test that pins today's answer so the day someone adds the field, it fails and points at them.

**Also in this commit, from my own re-review:**

- `cli_scan.py`'s `except`/`raise` pair was uncovered from the start — Codecov flagged it every round and I never acted. The API's `ValidationError` mapping got a test when it was added; its CLI counterpart didn't. That is precisely the asymmetry the cross-referencing comments I added one commit earlier claim to close. Now parametrized over all three exception types the handler names.
- Writing that test proved a comment from that same earlier commit wrong: it said `PackManifestError` is listed separately from `ValueError` "because it is not a `ValueError`". It **is** one, via `AbicheckError` — so the tuple was redundant, not necessary. Narrowed to `except ValueError`, which is simpler and strictly wider.

Local: 22711 passed, `ai-readiness` and `docs-contract` 0 errors, ruff/mypy clean, and both files' changed lines now measure fully covered.

---
_Generated by [Claude Code](https://claude.ai/code)_
